### ISSUE TITLE
Ember mug wont reconnect after resting HA from backup

### ISSUE THREAD

**Issue by rybackisback** [sopelj/hass-ember-mug-component#38]
* Home Assistant Version: 2023.11.2
* Operating System: 11.1
* Ember Mug Component Version: 0.8.0
* Intel NUC built in bluetooth

### Description

After restoring from backup ember mug does not connect to home assistant 

### What I Did

```
I reset the mug, I uninstalled and re-installed both hacs integration and ha  integration and attempted repair. I also ran bluetoothctl and put the mug into pairing mode but still it wont connect
```

### Logs / Errors

2023-11-17 20:52:43.351 DEBUG (MainThread) [ember_mug.mug] Set new device from DB:7C:3C:36:0C:88: Ember Ceramic Mug to DB:7C:3C:36:0C:88: Ember Ceramic Mug
2023-11-17 20:52:43.351 DEBUG (MainThread) [ember_mug.mug] Callback <bound method MugDataUpdateCoordinator._async_handle_callback of <custom_components.ember_mug.coordinator.MugDataUpdateCoordinator object at 0x7f040cea6010>> already registered
2023-11-17 20:52:43.351 DEBUG (MainThread) [custom_components.ember_mug.coordinator] Updating
2023-11-17 20:52:43.351 DEBUG (MainThread) [ember_mug.mug] Updating the following attributes: {'meta', 'firmware'}
2023-11-17 20:52:43.351 DEBUG (MainThread) [ember_mug.mug] Establishing a new connection from mug (ID: 139655508168464) to DB:7C:3C:36:0C:88: Ember Ceramic Mug
2023-11-17 20:52:44.375 DEBUG (MainThread) [ember_mug.mug] DB:7C:3C:36:0C:88: Ember Ceramic Mug: Failed to connect to the mug: (DB:7C:3C:36:0C:88) - DB:7C:3C:36:0C:88: Failed to connect after 8 attempt(s): [org.bluez.Error.Failed] Operation already in progress
2023-11-17 20:52:44.375 DEBUG (MainThread) [custom_components.ember_mug] Finished fetching ember-mug-db7c3c360c88 data in 12.542 seconds (success: False)
2023-11-17 20:52:44.375 DEBUG (MainThread) [ember_mug.mug] Unregistered callback: <function async_setup_entry.<locals>.<lambda> at 0x7f04093ec2c0>

```
```

**Comment by sopelj** [comment]
That's very strange. I have noticed Bluetooth Issues since upgrading to Home Assistant OS 11, is that what you are using? Bluetooth Proxies work fine, but it very rarely can connect via the USB dongle. I thought it might have been my dongle that was failing, but maybe there is another problem.

**Comment by rybackisback** [comment]
> That's very strange. I have noticed Bluetooth Issues since upgrading to Home Assistant OS 11, is that what you are using? Bluetooth Proxies work fine, but it very rarely can connect via the USB dongle. I thought it might have been my dongle that was failing, but maybe there is another problem.

Yes correct I am on is 11.1

**Comment by fowlermv** [comment]
I have the same issue. I moved my HA to new hardware and cannot get the ember to reconnect. Mine just says "failed setup, will retry". The mug is auto detected but never successfully adds into the restored HA instance. I've even gotten a new Bluetooth adapter since this and it didn't help. Any thoughts?

**Comment by sopelj** [comment]
Ok, that is very strange. I will do some investigation this week. Do either of you have errors on the host? I know, I am getting a bunch of kernel errors related to my Bluetooth adapter. I thought that was the cause. 

Are your mugs far from the adapter? I've noticed that it only seems to work when much closer now. 

Either this is an issue with Home Assistant OS 11 or it may be related to #36, where instead of retrying endlessly it now reports unavailable after a few tries. So, either it is a bit too lax or perhaps the former constant retrying was covering up issues with poor signal strength. It is strange however, that it doesn't seem to have this issue with Bluetooth Proxies, which is why I thought it might be the former. 

I'll try some tests this week.

**Comment by rybackisback** [comment]
I tried reconnecting 2 to 3m away from the BT adaptor and it is still the same.

**Comment by sopelj** [comment]
In my testing I noticed an official Bluetooth integration I use (the [Idasen Desk](https://rc.home-assistant.io/integrations/idasen_desk) also stopped working with my proxy disabled. 
There are a few different issues open in the core for it. I don't know if it's the same issue or not. 
The official documentation for that integration says:
> The recommended way to connect the desk to Home Assistant is by using an [ESPHome Bluetooth Proxy](https://esphome.io/projects/?type=bluetooth), as they provide the best experience. [...] The connection process is also less reliable than with ESPHome, so you may need to retry a few times until it succeeds. This is due to a limitation in the third-party Bluetooth implementation used by Home Assistant. Also, if not using an ESPHome Bluetooth Proxy, you may experience the desk becoming unavailable and needing re-pairing after a restart of Home Assistant or a reload of the integration

I'll spend some more time investigating and trying to fix this weekend, but in the meantime using Bluetooth Proxies is a good workaround and should offer a better experience regardless.

**Comment by sopelj** [comment]
I dug through the changelogs of all the bluetooth related commits and I can't find anything in 2023.11 that would account for this issue.

I did however manage to get the mug working again by going into bluetoothctl, removing the Mug and factory resetting it. After that it the integration re-connected it and it has been working stably since then. Does that work for anyone else?

The alternative is of course using a Proxy as they seem to have no problem at all.

**Comment by robert-cpl** [comment]
Have two mugs, Ember 1 and 2 versions. None of them managed to connect first try. Digging in the logs, I found the same issue as OP. I managed to fix my issue by, as sopelj suggested, to remove the mugs using `bluetoothctl` and retrying the integration.

**Comment by fowlermv** [comment]
I still cannot get it to work. I do think it has to do with the bluetooth in home assistant but I cannot remove the ember mug from my HA. I get "device MAC is not available" no matter what I do. It's listed in the device list but I cannot do anything about it.

**Comment by sopelj** [comment]
> I still cannot get it to work. I do think it has to do with the bluetooth in home assistant but I cannot remove the ember mug from my HA. I get "device MAC is not available" no matter what I do. It's listed in the device list but I cannot do anything about it.

Hi @fowlermv , have you tried removing the device from bluetoothctl and factory resetting it? 
Have you tried with a Bluetooth Proxy? Perhaps your Bluetooth adapter is problemtic. Does it work with other devices in Home Assistant?

**Comment by fowlermv** [comment]
My problem is that bluetoothctl does not work to actually remove the device. It has done this since restoring a backup while migrating from a virtual machine to hardware. All other bluetooth devices work fine though and I have tried two different bluetooth adapters which were both on the HA list of recommended hardware.

**Comment by sopelj** [comment]
> My problem is that bluetoothctl does not work to actually remove the device. It has done this since restoring a backup while migrating from a virtual machine to hardware. All other bluetooth devices work fine though and I have tried two different bluetooth adapters which were both on the HA list of recommended hardware.

Removing it does not work? How do you mean? Is it listed in the devices in bluetoothctl? If it never correctly connected that makes sense. Does it show up in `bluetoothctl` when in pairing mode if you set `scan on`? When using physical adpaters you might need to have the shell with `bluetoothctl` open in order for the device to pair. That is a limitation in the bluetooth library. Not sure why. Your other devices may not need to be paired. Most will jsut freely advertise their data.

**Comment by fowlermv** [comment]
Well I just decided to give it another shot today and it seems to have worked when previously it had not. We've been through a couple of HA updates though so maybe something got fixed somewhere.

**Comment by sopelj** [comment]
Awesome! I'm glad. I'll close this issue then, but don't hesitate if there are other issues.

**Comment by sopelj** [comment]
> I had a problem connecting a new mug, error "Failed setup, will retry: Could not find device with manufacturer data and address ...". It kept retrying without success.
> 
> What helped was:
> 1. Remove device from HA
> 2. `bluetoothctl remove <mac>`
> 3. Put cup into pairing mode
> 4. Now it was auto-detected again and I was able to add it

@puetsch that is a bit strange. It had never been connected before? Did you see anything in your error logs related to this? I'm glad you got it to work however. I'll do some tests and add that to the troubleshooting steps if there aren't any obvious issues to fix, thanks!

**Comment by puetsch** [comment]
I think what I did was that I started adding the mug, and only a second later or so put the cup into pairing mode.
Could that be why it ended up in a weird state where it was "bluetooth connected" but the ember integration wasn't able to initialize it? I also restarted HA while the cup was in pairing mode but it didn't help...

The cup had been paired with a phone before. And that phone was not in the house, so it couldn't have been connected to that.
